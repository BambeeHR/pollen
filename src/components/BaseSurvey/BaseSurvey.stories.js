import { action } from '@storybook/addon-actions';
import survey from '../../assets/surveys/HrAudit.json';
import onePageSurvey from '../../assets/surveys/HrAuditOnePage.json';
import BaseSurvey from './BaseSurvey.vue';

export default {
  title: 'BaseSurvey',
  component: BaseSurvey,
  parameters: {
    componentSubtitle: '',
  },
};

const handleInput = action('input');
const handleComplete = action('complete');

export const HrAudit = () => ({
  components: { BaseSurvey },
  data() {
    return { survey };
  },
  methods: {
    handleInput,
    handleComplete,
  },
  template: `
  <div class="max-w-2xl mx-auto my-4">
    <BaseSurvey
      :survey="survey"
      @input="handleInput"
      @complete="handleComplete"
    />
  </div>
  `,
});

export const HrAuditWithInitialValues = () => ({
  components: { BaseSurvey },
  data() {
    return {
      survey,
      initialValues: {
        hasPayrollSolution: false,
        hasPosters: 'no',
        hasPtoPolicy: 'yes',
        hasRemoteWorkers: false,
      },
    };
  },
  methods: {
    handleInput,
    handleComplete,
  },
  template: `
  <div class="max-w-2xl mx-auto my-4">
    <BaseSurvey
      :survey="survey"
      :initial-values="initialValues"
      @input="handleInput"
      @complete="handleComplete"
    />
  </div>
  `,
});

export const HrAuditOnePage = () => ({
  components: { BaseSurvey },
  data() {
    return { survey: onePageSurvey };
  },
  methods: {
    handleInput,
    handleComplete,
  },
  template: `
  <div class="max-w-2xl mx-auto my-4">
    <BaseSurvey
      :survey="survey"
      @input="handleInput"
      @complete="handleComplete"
    />
  </div>
  `,
});

export const HrAuditOnePageWithInitialValues = () => ({
  components: { BaseSurvey },
  data() {
    return {
      survey: onePageSurvey,
      initialValues: {
        hasPayrollSolution: false,
        hasPosters: 'no',
        hasPtoPolicy: 'yes',
        hasRemoteWorkers: false,
      },
    };
  },
  methods: {
    handleInput,
    handleComplete,
  },
  template: `
  <div class="max-w-2xl mx-auto my-4">
    <BaseSurvey
      :survey="survey"
      :initial-values="initialValues"
      @input="handleInput"
      @complete="handleComplete"
    />
  </div>
  `,
});
