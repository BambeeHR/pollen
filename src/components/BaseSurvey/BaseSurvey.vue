<template>
  <SurveyComponent :survey="model" class="base-survey" />
</template>

<script>
import cloneDeep from 'lodash/cloneDeep';
import * as SurveyVue from 'survey-vue';
import resolveConfig from 'tailwindcss/resolveConfig';
import tailwindConfig from '../../../tailwind.config';
import 'survey-vue/survey.css';

const { theme } = resolveConfig(tailwindConfig);

const SurveyComponent = SurveyVue.Survey;

export const mergeDefaultValues = (baseJson, defaultValues) => {
  const json = cloneDeep(baseJson);
  const elements = json.pages
    ? json.pages.map((page) => page.elements).flat()
    : json.elements;
  Object.entries(defaultValues).forEach(([key, value]) => {
    const matchingField = elements.find((el) => el.name === key);
    if (matchingField) {
      matchingField.defaultValue = value;
    }
  });
  return json;
};

export default {
  components: { SurveyComponent },
  props: {
    survey: {
      type: Object,
      required: true,
    },
    initialValues: {
      type: Object,
      default: () => ({}),
    },
  },
  data() {
    const json = mergeDefaultValues(this.survey, this.initialValues);
    const model = new SurveyVue.Model(json);
    return { model };
  },
  computed: {
    surveyData() {
      return this.model.data;
    },
  },
  watch: {
    surveyData() {
      this.$emit('input', this.model);
    },
  },
  created() {
    this.model.onComplete.add((sender) => {
      this.$emit('complete', sender);
    });

    const defaultThemeColors = SurveyVue.StylesManager.ThemeColors.default;
    defaultThemeColors['$border-color'] = theme.colors.gray['5'];
    defaultThemeColors['$body-container-background-color'] =
      theme.colors.gray['7'];
    defaultThemeColors['$disable-color'] = theme.colors.black;
    defaultThemeColors['$disabled-label-color'] = theme.colors.gray['3'];
    defaultThemeColors['$disabled-switch-color'] = 'rgba(72, 29, 115, 0.5)';
    defaultThemeColors['$disabled-slider-color'] = theme.colors.white;
    defaultThemeColors['$error-background-color'] = theme.colors['error-light'];
    defaultThemeColors['$error-color'] = theme.colors.error;
    defaultThemeColors['$inputs-background-color'] = theme.colors.white;
    defaultThemeColors['$main-color'] = theme.colors.primary;
    defaultThemeColors['$main-hover-color'] = theme.colors['primary-tint']['1'];
    defaultThemeColors['$slider-color'] = theme.colors.white;
    defaultThemeColors['$text-color'] = theme.colors.black;
    SurveyVue.StylesManager.applyTheme();

    // Other possible theme values to be declared:
    // $header-background-color: "#e7e7e7"
    // $header-color: "#6d7072"
    // $progress-text-color: "#9d9d9d"

    // Advance to the most recently completed page on the survey.
    const initialValueKeys = Object.keys(this.initialValues);
    if (
      initialValueKeys.length &&
      this.survey.pages.length > 1 &&
      !this.model.isCurrentPageHasErrors
    ) {
      let lastCompletedPageName;
      this.survey.pages.forEach((page) => {
        if (
          page.elements.some((element) =>
            initialValueKeys.includes(element.name)
          )
        ) {
          lastCompletedPageName = page.name;
        }
      });
      this.model.currentPage = lastCompletedPageName;
    }
  },
};
</script>

<style>
.base-survey.sv_main {
  @apply font-body-medium font-sans;
}

.base-survey.sv_main .sv_body {
  @apply border-none p-0;
}

.base-survey.sv_main input[type='button'],
.base-survey.sv_main button {
  @apply duration-ludicrous
    ease-out
    font-semibold
    px-4
    h-8
    rounded
    text-14
    transition-all;
}

.base-survey.sv_main textarea:hover {
  @apply border-gray-3;
}

.base-survey.sv_main .sv-boolean__switch {
  @apply duration-light
    ease-out 
    inline-flex 
    items-center 
    mx-2 
    transition-colors 
    w-12;
}

.base-survey.sv_main .sv-boolean--indeterminate .sv-boolean__switch {
  @apply bg-gray-4;
}

.base-survey.sv_main textarea:focus,
.base-survey.sv_main .sv-boolean input:focus ~ .sv-boolean__switch,
.base-survey.sv_main input[type='button']:focus,
.base-survey.sv_main button:focus {
  @apply outline-none shadow-outline;
}

.base-survey.sv_main .sv_container .sv_body .sv_p_root .sv_q textarea {
  @apply appearance-none border-2 font-body-text font-sans p-4 rounded-md;
}

.base-survey.sv_main
  .sv_q_erbox:not([style*='display: none']):not([style*='display:none']) {
  @apply bg-error-light border-none p-4 rounded-md text-error;
}

.base-survey.sv_main .sv-boolean__slider {
  @apply h-4 w-4;
}

.base-survey.sv_main .sv-boolean--indeterminate .sv-boolean__slider {
  margin-left: calc(50% - 8px);
}

.base-survey.sv_main .sv-boolean--checked .sv-boolean__slider {
  margin-left: calc(100% - 16px);
}

.base-survey.sv_main .sv_container .sv_body .sv_p_root .sv_q {
  @apply px-6 py-4;
}

.base-survey.sv_main input[type='radio'] {
  @apply sr-only;
}

.base-survey.sv_main .sv_q_radiogroup_label .circle {
  @apply align-middle
    border
    border-gray-4
    border-solid
    box-border
    h-6 
    inline-flex 
    items-center 
    justify-center 
    mr-4 
    relative 
    rounded-full 
    w-6;
}

.base-survey.sv_main .sv_q_radiogroup_label .circle svg {
  @apply hidden;
}

.base-survey.sv_main .checked .sv_q_radiogroup_label .circle {
  @apply border-2 border-primary;
}

.base-survey.sv_main .checked .sv_q_radiogroup_label .circle::before {
  @apply absolute bg-primary block h-0 rounded-full w-1/2;
  content: '';
  padding-top: 50%;
}
</style>
