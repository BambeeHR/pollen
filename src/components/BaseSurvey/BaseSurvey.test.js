import { mount } from '@vue/test-utils';
import get from 'lodash/get';
import BaseSurvey, { mergeDefaultValues } from './BaseSurvey.vue';
import survey from '../../assets/surveys/HrAudit.json';

describe('BaseSurvey', () => {
  describe('mergeDefaultValues', () => {
    test('applies default values to the survey model', () => {
      const updatedSurvey = mergeDefaultValues(survey, {
        hasRemoteWorkers: true,
      });
      expect(
        get(updatedSurvey, ['pages', 0, 'elements', 0, 'defaultValue'])
      ).toBe(true);
    });
  });
  test('renders correctly', () => {
    const wrapper = mount(BaseSurvey, { propsData: { survey } });
    expect(wrapper).toMatchSnapshot();
  });

  test('emits an input event when data changes', async () => {
    const wrapper = mount(BaseSurvey, { propsData: { survey } });
    await wrapper.find('input[name="hasRemoteWorkers"]').trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted().input[0]).toEqual([{ hasRemoteWorkers: true }]);
  });
  // TODO(jon.jandoc): Need to add test coverage for the complete event.

  test('advances to most recent page when passing initial values', async () => {
    const wrapper = mount(BaseSurvey, {
      propsData: {
        survey,
        initialValues: {
          hasPayrollSolution: false,
          hasPosters: 'no',
          hasPtoPolicy: 'yes',
          hasRemoteWorkers: false,
        },
      },
    });
    expect(
      wrapper.find('input[name="hasPayrollSolution"]').element
    ).toBeVisible();
  });
});
