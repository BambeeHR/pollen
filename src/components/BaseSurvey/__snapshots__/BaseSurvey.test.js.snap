// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseSurvey renders correctly 1`] = `
<div class="base-survey sv_main sv_default_css">
  <form onsubmit="return false;">
    <div class="sv_custom_header"></div>
    <div class="sv_container">
      <!---->
      <!---->
      <div class="sv_body">
        <!---->
        <!---->
        <!---->
        <div class="sv_p_root" id="sp_100">
          <!---->
          <!---->
          <div class="sv_row">
            <div class="sv-vue-row-additional-div">
              <div class="sv_q sv_qstn" id="sq_100" aria-labelledby="sq_100_ariaTitle" name="hasRemoteWorkers" style="flex-basis: 100.000000%; flex-grow: 1; flex-shrink: 1; width: 100.000000%; min-width: 300px; max-width: initial; display: inline-block;">
                <div class="">
                  <h5 aria-label="Do you have any remote workers?" id="sq_100_ariaTitle" class="sv_q_title">
                    <!----> <span class="sv_q_num" style="position: static;">1.</span>
                    <!----> <span style="position: static;"><span style="position: static;">Do you have any remote workers?</span></span> <span class="sv_q_required_text">*</span></h5>
                  <div class="sv_q_description"><span style="position: static;"><span style="position: static;"></span></span></div>
                </div>
                <div class="">
                  <div role="alert" id="sq_100_errors" class="sv_q_erbox sv_qstn_error_top" style="display: none;"></div>
                  <div class="sv_qcbc sv_qbln"><label class="sv-boolean sv-boolean--indeterminate"><input type="checkbox" name="hasRemoteWorkers" id="sq_100i" aria-required="true" aria-label="Do you have any remote workers?" class="sv-visuallyhidden" value=""> <span class="sv-boolean__label  ">No</span>
                      <div class="sv-boolean__switch"><span class="sv-boolean__slider"></span></div> <span class="sv-boolean__label  ">Yes</span>
                    </label></div>
                  <!---->
                  <!---->
                  <!---->
                </div>
                <!---->
              </div>
            </div>
          </div>
          <!---->
        </div>
        <!---->
        <div class="sv_progress">
          <div role="progressbar" aria-valuemin="0" aria-valuemax="100" class="sv_progress_bar" style="width: 10%;"><span class="undefined sv-hidden">Page 1 of 11</span></div> <span class="undefined undefined">Page 1 of 11</span>
        </div>
        <div class="sv_nav"><input type="button" value="Previous" class=" sv_prev_btn" style="display: none;"> <input type="button" value="Next" class=" sv_next_btn">
          <!----> <input type="button" value="Complete" class=" sv_complete_btn" style="display: none;"></div>
      </div>
      <!---->
      <!---->
      <!---->
      <!---->
    </div>
  </form>
</div>
`;
